libafl_qemu/src/qemu/qemu_opt.rs

```diff
@@ -0,0 +1,336 @@
+use core::fmt;
+use std::{fmt::Formatter, path::PathBuf};
+
+use strum_macros::Display;
+
+#[allow(non_camel_case_types)]
+#[derive(Default, Debug, Display, Clone, Copy)]
+pub enum QemuOptAccelerator {
+    kvm,
+    #[default]
+    tcg,
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Default, Debug, Display, Clone, Copy)]
+pub enum QemuOptDriveInterface {
+    floppy,
+    #[default]
+    ide,
+    mtd,
+    none,
+    pflash,
+    scsi,
+    sd,
+    virtio,
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Debug, Display, Clone, Copy)]
+pub enum QemuOptDiskImageFileFormat {
+    qcow2,
+    raw,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct QemuOptDrive {
+    file: Option<PathBuf>,
+    format: Option<QemuOptDiskImageFileFormat>,
+    interface: Option<QemuOptDriveInterface>,
+}
+
+impl fmt::Display for QemuOptDrive {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, " -drive")?;
+
+        let mut is_first_option = true;
+        let mut separator = || {
+            if is_first_option {
+                is_first_option = false;
+                " "
+            } else {
+                ","
+            }
+        };
+
+        if let Some(file) = &self.file {
+            write!(f, "{}file={}", separator(), file.to_str().unwrap())?;
+        }
+        if let Some(format) = &self.format {
+            write!(f, "{}format={}", separator(), format)?;
+        }
+        if let Some(interface) = &self.interface {
+            write!(f, "{}if={}", separator(), interface)?;
+        }
+
+        Ok(())
+    }
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Default, Debug, Display, Clone, Copy)]
+pub enum QemuOptSerial {
+    none,
+    null,
+    #[default]
+    stdio,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct QemuOpt {
+    accelerator: Option<QemuOptAccelerator>,
+    /// Set the directory for the BIOS, VGA BIOS and keymaps.
+    bios_path: Option<PathBuf>,
+    drives: Vec<QemuOptDrive>,
+    kernel: Option<PathBuf>,
+    load_vm: Option<PathBuf>,
+    machine: Option<String>,
+    monitor: Option<QemuOptSerial>,
+    no_graphic: bool,
+    /// ram size in MiB
+    ram_size: Option<u32>,
+    serial: Option<QemuOptSerial>,
+    smp_cpus: Option<u32>,
+    snapshot: bool,
+    vga_pci: bool,
+    do_not_start_cpu: bool,
+}
+
+impl fmt::Display for QemuOpt {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
```

Comment:
i think this is symptomatic of a design that could be improved.
if we add a new member in the struct above we could easily forget to extend this, the compiler will never complain if we forget something.
in qemu the full version of this struct is very large, it would become hard to maintain in the long run.

Reply:
I agree. In the new version the display impl is derived `libafl_derive::Display`. Each field of `QemuConfig` must implement the display trait (or must be a `Vec` or `Option` having an inner type implementing Display)

Reply:
solved now imho